### DISCUSSION TITLE
Outputting results from AssemblyRunner into an xml file.

### DISCUSSION THREAD

**Discussion by clicky** [xunit/xunit#2240] [Question] [answered]
Hey there,

I'm using XUnit in a bit of an odd way, but it's the only way to properly run tests for a project I am managing. We are writing a Plugin for [Rhino](https://www.rhino3d.com/) and I have 2 sets of tests. Real and Virtual. Virtual tests are ones that can be run headlessly no issue. However my real tests involve me opening the program, creating files and so on and so forth. To achieve this I've been using the AssemblyRunner class from Xunit.Runners. Whilst this works perfectly and outputs my tests nicely to the console application, I can't for the life of me figure out how to pipe these test results into a properly formatted xml file for my Build Pipelines. From what I can tell I need to format each TestFinishedInfo as well as the ExecutionCompleteInfo output somehow using the TransformFactory? Below is a stripped example of my code for simplicity.

	using System;
	using System.IO;
	using System.Linq;
	using System.Reflection;
	using System.Threading;
	using System.Collections.Generic;

	using Xunit;
	using Xunit.Runners;

	namespace MyApp.Tests.Real
	{

		public class Program
		{

			static object consoleLock = new object();

			static ManualResetEvent finished = new ManualResetEvent(false);

			public static void Main()
			{
				Assembly testassembly = Assembly.GetExecutingAssembly();

				using (var runner = AssemblyRunner.WithoutAppDomain(testassembly.Location))
				{
					runner.OnExecutionComplete = OnExecutionComplete;
					runner.OnTestFinished = OnTestFinished;

					Console.WriteLine("Discovering...");
					runner.Start();

					finished.WaitOne();
					finished.Dispose();
				}

			}

			static void OnExecutionComplete(ExecutionCompleteInfo info)
			{
				lock (consoleLock)
					Console.WriteLine($"Finished: {info.TotalTests} tests in {Math.Round(info.ExecutionTime, 3)}s ({info.TestsFailed} failed, {info.TestsSkipped} skipped)");

				finished.Set();
			}

			static void OnTestFinished(TestFinishedInfo info)
			{
				lock (consoleLock)
				{
					Console.WriteLine("[FINISHED] {0}: {1}", info.TestDisplayName, info.Output);
				}
			}
		}
	}

Any assistance would be welcomed 😊

**Comment by bradwilson** [comment, marked as answer]
At the moment we don't have any support for this with `AssemblyRunner`. You could do this with the full `XunitFrontController` but that's a lot more effort. I will be reworking the `AssemblyRunner` API for v3, so I'm not 100% sure where this would land in that.

**Reply by clicky** [reply]
Thanks @bradwilson, Sorry for the delay but I've been on other things. I've taken a look at XunitFrontController and so far I've managed to achieve what I had in AssemblyRunner. I've been having a poke around and it does seem more extensive, but I'm still not exactly sure how I push the results into an xml file, I assume I'd use the results from the MessageSink that's required for the RunAll methodm, somehow parse the instance of the XunitFrontController or even construct parts of it in each Visit method. Code sample below;

    public class Program
    {
        public static void Main(string[] args)
        {
            string assemblyPath = @"path";
            XunitFrontController controller = new XunitFrontController(
                assemblyPath);
    
            TestAssemblyConfiguration assemblyConfiguration = new TestAssemblyConfiguration();
    
            ITestFrameworkDiscoveryOptions discoveryOptions = TestFrameworkOptions.ForDiscovery(assemblyConfiguration);
            ITestFrameworkExecutionOptions executionOptions = TestFrameworkOptions.ForExecution(assemblyConfiguration);
    
            IMessageSink messageSink = new CustomTestMessageVisitor<ITestMessage>();
    
            Console.WriteLine("Running tests");
    
            controller.RunAll(
                messageSink: messageSink,
                discoveryOptions: discoveryOptions,
                executionOptions: executionOptions);
        }
    }
    
    public class CustomTestMessageVisitor<T> : TestMessageVisitor<T> where T : ITestMessage
    {
        protected override bool Visit(ITestFinished message)
        {
            Console.WriteLine("Test {0} finished.",
                message.Test.DisplayName);
            return true;
        }
    
        protected override bool Visit(ITestPassed message)
        {
            Console.WriteLine("Test {0} passed", message.Test.DisplayName);
            return true;
        }
    
        protected override bool Visit(ITestFailed message)
        {
            StringBuilder stringBuilder = new StringBuilder();
            foreach (string exceptionType in message.ExceptionTypes)
            {
                stringBuilder.AppendFormat("{0}, ", exceptionType);
            }
    
            Console.WriteLine("Test {0} failed with {1}", 
                message.Test.DisplayName, 
                stringBuilder.ToString());
            return true;
        }
    }

**Reply by bradwilson** [reply]
There is a `DelegatingXmlCreationSink` which we make part of an "execution sink chain", saving all the results into a pre-allocated `XElement` object.

Constructor:
https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.runner.utility/Sinks/DelegatingSinks/DelegatingXmlCreationSink.cs#L31-L42

Here we collect the XML:
https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/ConsoleRunner.cs#L382
https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/ConsoleRunner.cs#L439-L447

(You can see a couple others we use to support long-running test detection, turning skips into fails, and stashing the execution summaries so that we an print the final table of results.)

And here we deal with multi-assembly results including writing XML to disk [via transformers](https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/Utility/TransformFactory.cs#L21-L50):
https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/ConsoleRunner.cs#L328-L329
https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/ConsoleRunner.cs#L360

(The transformer for "xUnit.net v2" is just a [straight write to disk](https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/Utility/TransformFactory.cs#L61-L65); the other output formats [go through XSL-T stylesheets](https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/Utility/TransformFactory.cs#L80-L89) that we embed into the console runner.)

**Reply by clicky** [reply]
Thank you so much @bradwilson, I could not have asked for more. I'll digest this and see what I can do 😄

**Reply by clicky** [reply]
Good evening @bradwilson,
I've rifled through the code you sent and hacked around to see what I could get working, I'd really love to use the ConsoleRunner class but it seems that  it's not public so I can't create an instance of it. Which means I also can't use the ExecuteAssembly method inside of it that returns an XElement. I also searched for uses the final snippet you sent which is the bit I need to tap into, it seems that the [GetXmlTransformers ](https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.console/Utility/TransformFactory.cs#L56-L59) method from TransformFactory is only used in ConsoleRunner which makes me think that ConsoleRunner is my only way of accessing the XML creation, would I be correct in thinking that?

Kind Regards,

-- Callum

**Reply by bradwilson** [reply]
You are correct, there is no way to use this code via DLL today. You'll have to copy source to make it work.

**Reply by clicky** [reply]
And that would be the best way to run a TestSuite through xunit manually (Such as XUnitFrontController/AssemblyRunner or other) and output the results as XML?

**Reply by bradwilson** [reply]
Yes, for now, this is the best way. This will be easier in v3, which will include an API overhaul of AssemblyRunner.

**Reply by clicky** [reply]
Good to know! Thanks @bradwilson, now I can focus on getting this wrapped. Is there a timeline for v3's release?

**Reply by bradwilson** [reply]
Unfortunately, as you can tell by my delayed answer, v3 work doesn't have a fixed schedule (as it's a weekend side project for me) so I can't provide a date until we get closer.

**Reply by clicky** [reply]
Hello @bradwilson, is there any chance the latest v3 Release might have some of this functionality yet? :)

**Reply by bradwilson** [reply]
@clicketyclackety Not yet. I'm honestly not even sure AssemblyRunner is functioning properly yet, but it for sure will be getting at least evaluated for an architectural overhaul based on the new v3 runner APIs.

**Comment by ggeurts** [comment]
I have the same requirement of being able to trigger a test run ourselves and wanting to output the results to one or more formats that can be picked up by CI/CD environments.

**Reply by clicky** [reply]
I have managed to do this, but I end up with lots of problems with AppDomain etc. I need to try using System.Reflection to call upon the hidden classes I think.
